# Question 1

In [ ]:
import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import datasets
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import (train_test_split, cross_validate,
                                     GridSearchCV)
from sklearn.metrics import (accuracy_score, precision_score, recall_score,
                             confusion_matrix, PrecisionRecallDisplay)

In [ ]:
# Load data
X, y = datasets.fetch_openml('mnist_784', version=1, return_X_y=True,
                             as_frame=False)
X.shape

(70000, 784)

In [ ]:
# Take random subset of data for quicker processing
sample = np.random.choice(X.shape[0], 5000)
X = X[sample]
y = y[sample]

X.shape

(5000, 784)

In [ ]:
# Split data into train and test data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=1000)

In [ ]:
parameters = {
    'n_neighbors':range(1, 51)
}


knn = KNeighborsClassifier()
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)

grid = GridSearchCV(knn, param_grid=parameters, cv=10,
                    scoring='accuracy')
grid.fit(X_train, y_train)


GridSearchCV(cv=10, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': range(1, 51)}, scoring='accuracy')

In [ ]:
print("Best k value: ", grid.best_params_)
print("Best score: ", grid.score(X_test, y_test))

Best k value:  {'n_neighbors': 1}
Best score:  0.927


# Question 2

The implemented model is able to cross-validate the chosen value using a 10-fold approach (cv = 10). The performance of this model can be characterized by the accuracy score of the best k value, which represents the percentage of correct classifications by the model with that particular k value. The model with the automated best-fit value of k=1 produces a 0.927 accuracy score. By manually generating a random value of k in the final cell of question 1, it can be seen that the other values of k in the given range of numbers are lower than the ideal value found by the grid search model that was implemented. As such, this model performs well based on accuracy. 

In [ ]:
import random as rd

k = rd.randint(1, 51)
print(f"Value of K is: {k}")

knn = KNeighborsClassifier(n_neighbors = k)
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)

print("Score is : ", accuracy_score(y_test, y_pred))

Value of K is: 23
Score is :  0.9


# Question 3

In theory, the model should generalize well to new datasets that are similar since the MNIST data set contains many datapoints based on which the model can be trained. To better fit the data to a new dataset, utilizing a larger sample size than the one used in this model for the sake of performance can be beneficial. By training the data using a larger sample, the accuracy, precision and recall of the model can be enhanced and be better suited to be applied to new datasets. 

# Question 4

The purpose of the narrow AI is to learn patterns in handwritten digits to corectly identify similar digits. The AI creates predictions based on a given image and classifies it to an integer. It provides accuracy scores that measure the percentage of correclty classifications by the model. 

# Question 5

A major example that comes to mind is the use of AI & drone technology in warfare. The underlying image classification AI being used to identify of targets is similar to the model produced in class using the MNIST, although at a much more advanced level. The essential purpose of this AI is to provide classification of an object based on the model it has been trained on. The creator of such a classification model could have intended to use it to identify cancerous cells, enhance autopilot/cruise control capabilities in cars to improve safety, or to even make major ehancements in the robotics field.
Instead, the AI is being used as a weapon of war rather than an instrument of good. The objective of the AI is simple image classification, however the actions of the AI do not necessarily benefit society since the AI is being used as a weapon of mass destruction.